src/backend/executor/execUtils.c

```diff
@@ -1536,6 +1539,22 @@ void mppExecutorFinishup(QueryDesc *queryDesc)
 		ExecSquelchNode(queryDesc->planstate);
 	}
 
+	/*
+	 * Notify corresponding producer that the reading is completed. We need to
+	 * do notification before processing the results in order to avoid a
+	 * deadlock. Deadlock could occur when the consumer of a cross slice Shared
+	 * Scan was in slice 0 and the producer was in another slice. QE with
+	 * producer couldn't complete until the consumer sent the notification,
+	 * however that didn't happen because consumer at slice 0 send a
+	 * notification after all QEs are completed.
+	 */
+	ListCell   *cell;
+	foreach (cell, estate->sharedScanConsumers)
+	{
+		ShareInputScanState *state = lfirst(cell);
+		shareinput_reader_notifydone(state);
```

Comment:
why not something like
```suggestion
		shareinput_local_state *local_state = state->local_state;
		shareinput_Xslice_reference *ref = state->ref;

		if (ref == NULL || local_state->closed)
			continue;

		ShareInputScan *sisc = (ShareInputScan *) state->ss.ps.plan;
		shareinput_reader_notifydone(ref, sisc->nconsumers);
		local_state->closed = true;
```

Reply:
`shareinput_Xslice_reference *ref` is defined at the `nodeShareInputScan.c` file and we can not use it at the other files.

Reply:
`struct shareinput_local_state` is the same.

Reply:
maybe then create public wrapper?